libexec/charliecloud.py

```diff
@@ -0,0 +1,637 @@
+                "path",
+                "name",
+                "tag",
+                "digest")
+
+   def __init__(self, src=None):
+      self.host = None
+      self.port = None
+      self.path = []
+      self.name = None
+      self.tag = None
+      self.digest = None
+      if (isinstance(src, str)):
+         src = self.parse(src)
+      if (isinstance(src, lark.tree.Tree)):
+         self.from_tree(src)
+      elif (src is not None):
+         assert False, "unsupported initialization type"
+
+   def __str__(self):
+      out = ""
+      if (self.host is not None):
+         out += self.host
+      if (self.port is not None):
+         out += ":" + str(self.port)
+      if (self.host is not None):
+         out += "/"
+      out += self.path_full
+      if (self.tag is not None):
+         out += ":" + self.tag
+      if (self.digest is not None):
+         out += "@sha256:" + self.digest
+      return out
+
+   @staticmethod
+   def parse(s):
+      if ("%" in s):
+         s = s.replace("%", "/")
+      parser = lark.Lark("?start: image_ref\n" + GRAMMAR, parser="earley",
+                         propagate_positions=True)
+      try:
+         tree = parser.parse(s)
+      except lark.exceptions.UnexpectedInput as x:
+         FATAL("image ref syntax, char %d: %s" % (x.column, s))
+      except lark.exceptions.UnexpectedEOF as x:
+         # We get UnexpectedEOF because of Lark issue #237. This exception
+         # doesn't have a column location.
+         FATAL("image ref syntax, at end: %s" % s)
+      DEBUG(tree.pretty(), v=2)
+      return tree
+
+   @property
+   def as_verbose_str(self):
+      def fmt(x):
+         if (x is None):
+            return None
+         else:
+            return repr(x)
+      return """\
+as string:    %s
+for filename: %s
+fields:
+  host    %s
+  port    %s
+  path    %s
+  name    %s
+  tag     %s
+  digest  %s\
+""" % tuple(  [str(self), self.for_path]
+            + [fmt(i) for i in (self.host, self.port, self.path,
+                                self.name, self.tag, self.digest)])
+
+   @property
+   def for_path(self):
+      return str(self).replace("/", "%")
+
+   @property
+   def path_full(self):
+      out = ""
+      if (len(self.path) > 0):
+         out += "/".join(self.path) + "/"
+      out += self.name
+      return out
+
+   @property
+   def version(self):
+      if (self.tag is not None):
+         return self.tag
+      if (self.digest is not None):
+         return "sha256:" + self.digest
+      assert False, "version invalid with no tag or digest"
+
+   @property
+   def url(self):
+      out = ""
+      return out
+
+   def copy(self):
+      "Return an independent copy of myself."
+      return copy.deepcopy(self)
+
+   def defaults_add(self):
+      "Set defaults for all empty fields."
+      if (self.host is None): self.host = "registry-1.docker.io"
+      if (self.port is None): self.port = 443
+      if (len(self.path) == 0): self.path = ["library"]
```

Comment:
Should we make this default `aarch64` is the host is aarch64? Otherwise all the images pulled won't work. Here is the list of images available: https://hub.docker.com/search?q=aarch64&type=image.

Reply:
Yes, but let's push that to a future PR.